### PR TITLE
Fixed #5927 - Now remembers page rotation on reload

### LINKED ISSUE #5927: Viewer doesn't remember last rotation angle
The viewer does remember last scroll position but not rotation angle. After reload the viewer will restore the last scroll position but because the rotation angle is not restored, the page position may be wrong.

### PR DESCRIPTION
**Pull request by mbbaig**
Fix for [#5927](https://github.com/mozilla/pdf.js/issues/5927)

### CODE REVIEW

File: web/viewer.js [hunk 1]
```diff
@@ -1786,7 +1789,8 @@ window.addEventListener('updateviewarea', function () {
       'page': location.pageNumber,
       'zoom': location.scale,
       'scrollLeft': location.left,
-      'scrollTop': location.top
+      'scrollTop': location.top,
+      'pageRotation': PDFViewerApplication.pdfViewer.pagesRotation
```
Comment (Snuffleupagus) [inline web/viewer.js]: I think you should add `rotation: this._pagesRotation` at [pdf_viewer.js#L572](https://github.com/mozilla/pdf.js/blob/master/web/pdf_viewer.js#L572) instead, and change this line to `'rotation': location.rotation`.

**Comment by Snuffleupagus** [comment]
When testing this locally, and in the preview, the following is printed in the console:
```
"pdfViewSetScale: '0' is an unknown zoom value." viewer.js:4483:1
```

File: web/viewer.js [hunk 2]
```diff
@@ -902,6 +902,9 @@ var PDFViewerApplication = {
           var left = store.get('scrollLeft', '0');
           var top = store.get('scrollTop', '0');
 
+          self.pageRotation = store.get('pageRotation', '0');
+          self.pdfViewer.pagesRotation = self.pageRotation;
```
Comment (Snuffleupagus) [inline web/viewer.js]: Unfortunately this won't work, since it breaks the thumbnails (and it's probably also related to the console error).
I believe that you need to use `PDFViewerApplication.rotatePages` instead, see [viewer.js#L1326](https://github.com/mozilla/pdf.js/blob/master/web/viewer.js#L1326).

**Comment by Snuffleupagus** [comment]
A more general question: I'm well aware that rotation isn't specified in the ["Open parameters"](http://www.adobe.com/content/dam/Adobe/en/devnet/acrobat/pdfs/pdf_open_parameters.pdf), but do we want to expose it as a hash parameter?

**Comment by mbbaig** [comment]
I made all of the changes mentioned.

I changed the ```PDFViewerApplication.rotatePages``` function to the following:
```javascript
rotatePages: function pdfViewRotatePages(delta) {
    this.pageRotation = (this.pageRotation + 360 + delta) % 360;
    this.pdfViewer.pagesRotation = this.pageRotation;
    this.pdfThumbnailViewer.pagesRotation = this.pageRotation;

    this.forceRendering();
  }
```
As for the error. It occurs because ```this._currentScaleValue``` is set to '0'. Here is the location [pdf_viewer.js#L201](https://github.com/mozilla/pdf.js/blob/master/web/pdf_viewer.js#L201)

I can't see where that is being set to '0'.

**Comment by mbbaig** [comment]
@Snuffleupagus I think exposing the rotation as a parameter couldn't hurt. It would also make the code consistent here [viewer.js#905](https://github.com/mozilla/pdf.js/blob/master/web/viewer.js#L905). We could simply add rotation into ```storedHash```

File: web/pdf_viewer.js [hunk 3]
```diff
@@ -197,7 +197,6 @@ var PDFViewer = (function pdfViewer() {
         var page = this.pages[i];
         page.update(page.scale, rotation);
       }
-
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js]: This change is unrelated to the patch, please revert it.

**Comment by Snuffleupagus** [comment]
> I changed the PDFViewerApplication.rotatePages function to the following:
I'm sorry, but changing it in this way will first of all break other functionality, and second of all not actually achieve the desired goal! What you should do is e.g. something like this:
```javascript
rotatePages: function pdfViewRotatePages(delta, noRendering) {
  this.pageRotation = (this.pageRotation + 360 + delta) % 360;
  this.pdfViewer.pagesRotation = this.pageRotation;
  this.pdfThumbnailViewer.pagesRotation = this.pageRotation;

  if (!noRendering) {
    this.forceRendering();

    this.pdfViewer.scrollPageIntoView(pageNumber);
  }
},
```
Then you can call `self.rotatePage(rotation, true)` when you've read the parameter from storage.

**Comment by mbbaig** [comment]
@Snuffleupagus updated to code to your specifications moved to code to after init so no more error.

**Comment by Snuffleupagus** [comment]
> moved to code to after init so no more error.

Unfortunately I don't think that is going to work correctly either, since I believe that you need to set the rotation *before* calling `setInitialView`, but in such a way that it doesn't trigger the error.

File: web/viewer.js [hunk 4]
```diff
@@ -901,6 +902,12 @@ var PDFViewerApplication = {
                      store.get('zoom', self.pdfViewer.currentScale);
           var left = store.get('scrollLeft', '0');
           var top = store.get('scrollTop', '0');
+          rotation = store.get('rotation', '0');
+
+          self.setScale(DEFAULT_SCALE, true);
```
Comment (mbbaig) [inline web/viewer.js]: I am now setting the scale and moved the code back to its original spot. There is no error. Can this be considered a viable solution?

Comment (Snuffleupagus) [inline web/viewer.js]: > Can this be considered a viable solution?

In my opinion, no. This patch tries to add a minor feature, hence I don't think that unconditionally triggering an unnecessary rendering operation during *every* document load is a particularly good thing to do.

Given that the error isn't really a breaking error, but just a console message, I'd *remove* this line and try something else instead: Try changing https://github.com/mozilla/pdf.js/blob/master/web/pdf_viewer.js#L201 to
```javascript
if (this._currentScale !== UNKNOWN_SCALE)
  this._setScale(this._currentScaleValue, true);
}
```

*If this seem to work: please make sure that you test the patch __thoroughly__, to ensure first of all that it does not regress other functionality, and second of all that the rotation works as intended!*

File: web/viewer.js [hunk 5]
```diff
@@ -894,6 +894,7 @@ var PDFViewerApplication = {
 
       store.initializedPromise.then(function resolved() {
         var storedHash = null;
+        var rotation = 0;
```
Comment (Snuffleupagus) [inline web/viewer.js]: Please remove this line.

File: web/viewer.js [hunk 6]
```diff
@@ -901,6 +902,12 @@ var PDFViewerApplication = {
                      store.get('zoom', self.pdfViewer.currentScale);
           var left = store.get('scrollLeft', '0');
           var top = store.get('scrollTop', '0');
+          rotation = store.get('rotation', '0');
```
Comment (Snuffleupagus) [inline web/viewer.js]: You should ensure that `rotation` is an integer. Change the line to
`var rotation = store.get('rotation', '0') | 0;` in order to do that.

File: web/viewer.js [hunk 7]
```diff
@@ -901,6 +902,12 @@ var PDFViewerApplication = {
...
+
+          self.setScale(DEFAULT_SCALE, true);
+          if (rotation !== self.pageRotation) {
+            self.rotatePages(rotation,true);
```
Comment (Snuffleupagus) [inline web/viewer.js]: Nit: add a `space` after the comma.

Comment (Snuffleupagus) [inline web/viewer.js]: Nit: add a space after the comma.

**Comment by mbbaig** [comment]
updated and tested. seems to work fine with no errors

File: web/viewer.js [hunk 8]
```diff
@@ -901,6 +901,11 @@ var PDFViewerApplication = {
                      store.get('zoom', self.pdfViewer.currentScale);
           var left = store.get('scrollLeft', '0');
           var top = store.get('scrollTop', '0');
+          var rotation = store.get('rotation', '0') | 0;
+
+          if (rotation !== self.pageRotation) {
+            self.rotatePages(rotation,true);
```
Comment (timvandermeij) [inline web/viewer.js]: Nit: add a space after the `,`, i.e., `self.rotatePages(rotation, true)`

**Comment by xlc** [comment]
@Snuffleupagus Can you have a look and maybe merge this?

File: web/pdf_viewer.js [hunk 9]
```diff
@@ -198,7 +198,9 @@ var PDFViewer = (function pdfViewer() {
         page.update(page.scale, rotation);
       }
 
-      this._setScale(this._currentScaleValue, true);
+      if (this._currentScale !== UNKNOWN_SCALE) {
```
Comment (brendandahl) [inline web/pdf_viewer.js]: Why is this change needed?

File: web/viewer.js [hunk 10]
```diff
@@ -901,6 +901,11 @@ var PDFViewerApplication = {
                      store.get('zoom', self.pdfViewer.currentScale);
           var left = store.get('scrollLeft', '0');
           var top = store.get('scrollTop', '0');
+          var rotation = store.get('rotation', '0') | 0;
+
+          if (rotation !== self.pageRotation) {
```
Comment (brendandahl) [inline web/viewer.js]: The code has changed here a bit. Instead of calling this here, let's pass the rotation as an argument into setInitialView

File: web/viewer.js [hunk 11]
```diff
@@ -1323,15 +1328,16 @@ var PDFViewerApplication = {
...
+  rotatePages: function pdfViewRotatePages(delta, noRendering) {
     var pageNumber = this.page;
     this.pageRotation = (this.pageRotation + 360 + delta) % 360;
     this.pdfViewer.pagesRotation = this.pageRotation;
     this.pdfThumbnailViewer.pagesRotation = this.pageRotation;
 
-    this.forceRendering();
-
-    this.pdfViewer.scrollPageIntoView(pageNumber);
+    if (!noRendering) {
```
Comment (brendandahl) [inline web/viewer.js]: I like to avoid double negatives when possible. How about we make a setRotation(delta) function that only updates values and then change rotatePages to call setRotation and the code in this 'if' statement. Then in setInitialView you just have to call setRotation(...)

**Comment by brendandahl** [comment]
Needs a rebase and comments addressed.

**Comment by timvandermeij** [comment]
Closing in favor of #6980.
